prometheus_juju_backup_all_exporter/metrics.py

```diff
@@ -30,105 +33,127 @@ def _process(self, data, old_sample={}):
     """
 
     @abstractmethod
-    def _process(self, data, old_sample):
+    def _process(
+        self, data: Dict[str, List[Dict[str, Any]]], old_sample: Dict[Tuple, float]
+    ) -> List[Dict[str, Any]]:
         """User defined process method used by `self.add_samples()`.
 
         Args:
             data: the data to be processed
             old_sample: previous metric sample, useful when implementing counter.
 
         Returns:
-            processed data: list of dictionary containing 'labels' (list of str) and 'value' (float).
+            processed data: list of dictionary containing 'labels' (list of
+            str) and 'value' (float).
 
         Examples:
-            returned_processed_data = [{"label": [], "value": 10.0}, {"label": ["foo"], "value": 20.0}]
+            returned_processed_data = [
+                {"label": [], "value": 10.0}, {"label": ["foo"], "value": 20.0}
+            ]
         """
-        pass  # pragma: no cover
+        return []  # pragma: no cover
 
     @property
-    def previous_data(self):
+    def previous_data(self) -> Dict[Tuple, float]:
         """Return previous sample as a labels <-> value pair."""
-        return {tuple(sample.labels.values()): sample.value for sample in self.samples}
+        return {
+            tuple(sample.labels.values()): sample.value
+            for sample in self.samples  # type: ignore[attr-defined] # pylint: disable=E1101
```

Comment:
I understand that this abstract class is supposed to be used only as Mixin with specific type of class that has attributes like `samples` and `name`. However wouldn't it be better if this class would define these attributes (with empty values) to avoid `pylint E1101`?

Reply:
See above comment